bdsg/include/bdsg/snarl_distance_index.hpp

```diff
@@ -822,176 +999,117 @@ class SnarlDistanceIndex : public SnarlDecomposition, public TriviallySerializab
 
-        virtual bool has_connectivity(connectivity_t connectivity) const;
-        virtual bool has_connectivity(endpoint_t start, endpoint_t end);
+        bool has_connectivity(connectivity_t connectivity) const;
+        bool has_connectivity(endpoint_t start, endpoint_t end);
 
         //Get and set a pointer to this node's parent, including its orientation
-        //TODO: I don't think it matters if a chain is reversed or not, also it might not matter if a snarl is
-        virtual size_t get_parent_record_offset() const;
+        size_t get_parent_record_offset() const;
 
         //Get and set the minimum length (distance from start to end, including boundaries for 
         // chains but not snarls, just node length for nodes)
-        virtual size_t get_min_length() const;
+        size_t get_min_length() const;
         
         //Get and set this node's maximum length
         //This isn't actually a maximum, it's the maximum among minimum distance paths 
         //through each node in the snarl/chain
-        virtual size_t get_max_length() const;
+        size_t get_max_length() const;
 
         //Get and set this structure's rank in its parent
         //For children of snarls, this means the actual rank
         //For children of chains, it points to the node in the chain
-        virtual size_t get_rank_in_parent() const;
+        size_t get_rank_in_parent() const;
 
         //Is this node reversed in its parent
-        //TODO: I think a node is the only thing that can be reversed in its parent, and only if it is in a chain
-        virtual bool get_is_reversed_in_parent() const;
+        bool get_is_reversed_in_parent() const;
 
         //Get the node id of the start/end of this structure (start node of a snarl/chain)
-        //TODO: Also need to add min node id
-        virtual handlegraph::nid_t get_start_id() const; 
+        handlegraph::nid_t get_start_id() const; 
         //True if the node is traversed backwards to enter the structure
-        virtual bool get_start_orientation() const;
-        virtual handlegraph::nid_t get_end_id() const;
+        bool get_start_orientation() const;
+        handlegraph::nid_t get_end_id() const;
         //Return true if the end node is traversed backwards to leave the snarl
-        virtual handlegraph::nid_t get_end_orientation() const;
+        handlegraph::nid_t get_end_orientation() const;
 
-        //TODO: These are redeclared so that I don't need to pass the SnarlTreeRecord the actual distance index
-        //Get the offset into snarl_tree_records for a node record
-        virtual size_t get_offset_from_id (const handlegraph::nid_t id) const;
     };
 
     //Record interpreter that has a non-const reference to snarl_tree_records, so it can
     //also add things
     struct SnarlTreeRecordConstructor {
 
-        //TODO: This might be bad but this has the same members as SnarlTreeRecord
```

Comment:
I think the next line doesn't make sense without this one.

You do sometimes have to duplicate code to deal with const-ness. I don't think you can inherit from something and make a field it had that was const not const anymore. Though you could keep it const and do a bunch of const casting.

You could try templating things on the pointer type. So you have:

```

template<typename P>
struct BaseSnarlTreeRecord {
    P records;
    size_t offset;
    ...
};

using SnarlTreeRecord = BaseSnarlTreeRecord<const bdsg::yomo::UniqueMappedPointer<bdsg::MappedIntVector>*>;
struct MutableSnarlTreeRecord: public BaseSnarlTreeRecord<bdsg::yomo::UniqueMappedPointer<bdsg::MappedIntVector>*> {
...
};
```